### PR TITLE
i3, sway: replace fonts with submodule

### LINKED ISSUE #1937: i3 / sway font configuration is misleading
#### Font size placement

[The description and example](https://github.com/nix-community/home-manager/blob/6aa6556bcab6dc0f6398b4daa8404d788fd7a6a2/modules/services/window-managers/i3-sway/lib/options.nix#L13-L17) for the `i3.config.fonts` (respectively `sway`) option is misleading. I just spent 30 minutes debugging a font display issue before opening [the i3 documentation](https://i3wm.org/docs/userguide.html#fonts) and realizing the example and description were wrong :smile: .

Specifically, the i3 documentation example shows that **only one** font size can be passed in the entire line, after a comma separated list of font families, so the following example:

```nix
[ "FontAwesome 10" "Terminus 10" ]
```

should be

```nix
["FontAwesome" "Terminus 10" ]
```

to require the FontAwesome and Terminus fonts, in size 10.

Note that this makes the semantics of the `fonts` option weird, as only the last font in the list should have a size.

How about introducing `fontSize` as an additional option, and append it at the end of the line, to guarantee correctness? This would break backwards compatibility though, unless it's not appended when unset.

#### Description is wrong about font ordering

Additionally, the description states:

> The order here is important (e.g. icons font should go before the one used for text).

While it's true that the order is important, icons fonts should go **after** the one used for text, at least in my testing. Putting FontAwesome first leads i3 to use it for text as well, and it's not a monospace font.

### Illustration of the font size issue

1. With this in my configuration, and font-awesome installed from nixpkgs, I get the correct behavior:

```nix
{
  bars = [
    {
      statusCommand = "i3status-rs ${barConfigPath}";
      fonts = [ "DejaVuSansMono" "FontAwesome5Free 9" ];
    }
  ];
}
```

image

Note the correct volume muted icon.

2. With this configuration:

```nix
{
  bars = [
    {
      statusCommand = "i3status-rs ${barConfigPath}";
      fonts = [ "DejaVuSansMono 9" "FontAwesome5Free 9" ];
    }
  ];
}
```

I get this:

image

The font isn't DejaVuSansMono, probably because i3 parsed the font line `DejaVuSansMono 9, FontAwesome5Free 9` differently and didn't find a match for `DejaVuSansMono 9`. The icon is still fine though.

3. Finally using this configuration:

```nix
{
  bars = [
    {
      statusCommand = "i3status-rs ${barConfigPath}";
      fonts = [ "FontAwesome5Free 9" "DejaVuSansMono 9" ];
    }
  ];
}
```

I get 
image

Note the broken volume mute icon, because this time it's FontAwesome that isn't found correctly because of the extra font size added.

#### Maintainer CC

@alexarice @sumnerevans for sway, couldn't find any maintainer for i3wm.

#### Technical details

```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.27, NixOS, 21.05.20210410.a73020b (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20210326_dd77f71`
 - channels(root): `"nixos-21.05pre283367.0a5f5bab0e0"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by sumnerevans**
### Description

This applies to both the root-level and the bar configs.

Closes #1937.

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/doc/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

**Comment by sumnerevans** [comment]
I'm not sure how to deprecate the old config options. I tried adding the following in the `i3.nix` file:

```
  imports = [
    (mkRemovedOptionModule [
      "xsession"
      "windowManager"
      "i3"
      "config"
      "fonts"
    ] ''
      Fonts are now configured through the 

        xsession.windowManager.i3.config.fontNames

      and

        xsession.windowManager.i3.config.fontSize

      options. Please use those instead.
    '')
  ];
```

but when I try and build using

```
home-manager -I home-manager=$HOME/projects/home-manager switch
```

I get
```
error: The option `wayland.windowManager.sway.config' in `/home/sumner/projects/home-manager/modules/services/window-managers/i3-sway/sway.nix' is a prefix of options in `/home/sumner/projects/home-manager/modules/services/window-managers/i3-sway/sway.nix'.
(use '--show-trace' to show detailed location information)
```

**Comment by sumnerevans** [comment]
@alarsyo here's the PR for #1937. I don't have permission to add you as a reviewer.

File: tests/modules/services/window-managers/i3/i3-followmouse-expected.conf
Comment (alarsyo) [inline tests/modules/services/window-managers/i3/i3-followmouse-expected.conf:1]: I'd say that looks a bit weird, is there a way to specify float precision when doing string interpolation with them? Although if i3 and sway parse these correctly, it's probably fine :)

**Review by alarsyo — COMMENTED** [review]
That looks great, thanks! Do you think the font styles also deserve a separate setting, or putting those in the last fontName should be enough?

https://github.com/nix-community/home-manager/issues/1937#issuecomment-822832741

File: tests/modules/services/window-managers/i3/i3-followmouse-expected.conf
Comment (sumnerevans) [reply tests/modules/services/window-managers/i3/i3-followmouse-expected.conf:1]: Both i3 and sway parse these correctly. I think this is the best approach because it allows for fractional font size specification.

**Comment by sumnerevans** [comment]
@alarsyo oh, I think it does make sense to add `fontStyle` as well. I missed that comment on the original issue.

File: modules/services/window-managers/i3-sway/lib/functions.nix
Comment (alarsyo) [inline modules/services/window-managers/i3-sway/lib/functions.nix:42]: nitpick: the default `""` for `style` isn't needed if the corresponding option already has that default value, I think?

**Review by alexarice — APPROVED** [review]
Looks reasonable to me

**Comment by sumnerevans** [comment]
@alexarice do you have any ideas for how to deprecate the options?

**Comment by alexarice** [comment]
Ah right, that is a good point, the change should probably be hidden behind a `stateVersion` check

**Comment by sumnerevans** [comment]
I'm not sure that we need to gate this behind a `stateVersion`, but we definitely need some deprecation warnings. Maybe @rycee or @berbiche can advise?

**Comment by alexarice** [comment]
I thought anything that can break configs should be behind a `stateVersion` change

File: tests/modules/services/window-managers/sway/sway-null-package.conf
Comment (teto) [inline tests/modules/services/window-managers/sway/sway-null-package.conf:1]: this isnt right, it should remain `8`

File: modules/services/window-managers/i3-sway/lib/options.nix
Comment (teto) [inline modules/services/window-managers/i3-sway/lib/options.nix:94]: would be nice if we could reuse types.fontType

File: tests/modules/services/window-managers/sway/sway-null-package.conf
Comment (sumnerevans) [reply tests/modules/services/window-managers/sway/sway-null-package.conf:1]: see discussion here: https://github.com/nix-community/home-manager/pull/1950#discussion_r619927313

File: modules/services/window-managers/i3-sway/lib/options.nix
Comment (sumnerevans) [reply modules/services/window-managers/i3-sway/lib/options.nix:94]: The problem is that this requires a list of fonts, not a single font.

File: tests/modules/services/window-managers/sway/sway-null-package.conf
Comment (teto) [reply tests/modules/services/window-managers/sway/sway-null-package.conf:1]: ha sry had not seen.

File: modules/services/window-managers/i3-sway/lib/options.nix
Comment (teto) [reply modules/services/window-managers/i3-sway/lib/options.nix:94]: wont a `listOf fontType` work ?
https://i3wm.org/docs/userguide.html#fonts

Comment (sumnerevans) [reply modules/services/window-managers/i3-sway/lib/options.nix:94]: The issue is that there can only be a single font size and style specified (that was the original issue described in #1937) for all of the font names.

File: tests/modules/services/window-managers/sway/sway-null-package.conf
Comment (sumnerevans) [reply tests/modules/services/window-managers/sway/sway-null-package.conf:1]: no worries :) thanks for the review!

File: modules/services/window-managers/i3-sway/lib/options.nix
Comment (teto) [reply modules/services/window-managers/i3-sway/lib/options.nix:94]: maybe add a comment about fractional scaling, I think it may be surprising for people (it was for me).
Also this is a breaking change why rename the options ? `fonts` was fine, I get `The option `home-manager.users.teto.xsession.windowManager.i3.config.fonts' does not exist.`. In fact I feel like this may be overengineered, we dont need the complexity, we can just fix the example. Even with the change, people can still shoot themselves in the foot. The PR might be useful if someone merges different configurations for the font option but that's really an edgecase I feel like.

Comment (sumnerevans) [reply modules/services/window-managers/i3-sway/lib/options.nix]: I agree that people can still shoot themselves in the foot, but I think it will be a lot harder to do so since it is more obvious how you are supposed to use it (as opposed to before where you had to have "insider information" about how both the hm module works and how i3/sway expects fonts to be specified).

I've gone ahead and renamed to `fonts` and added backwards compatibility for the old list-only version of specifying the font.

File: modules/services/window-managers/i3-sway/lib/options.nix
Comment (teto) [inline modules/services/window-managers/i3-sway/lib/options.nix:28]: where can I find these styles ? is this a pango property ?

**Review by teto — APPROVED** [review]
Looks good here. I'll let some time for other to comment and we can merge in a day or two.

File: modules/services/window-managers/i3-sway/lib/options.nix
Comment (sumnerevans) [reply modules/services/window-managers/i3-sway/lib/options.nix:28]: Yes, it's part of pango. Unfortunately, in my bit of searching, I can't find any reference to the actual set of styles that you can apply.

File: modules/services/window-managers/i3-sway/lib/options.nix
Comment (berbiche) [inline modules/services/window-managers/i3-sway/lib/options.nix:14]:
```suggestion
        default = [ "monospace" ];
        defaultText = literalExample ''[ "monospace" ]'';
```

File: modules/services/window-managers/i3-sway/lib/options.nix
Comment (berbiche) [inline modules/services/window-managers/i3-sway/lib/options.nix:19]:
```suggestion
        example = literalExample ''[ "FontAwesome" "Terminus" ]'';
```

**Review by berbiche — APPROVED** [review]
Some documentation changes required, otherwise LGTM!

**Comment by sumnerevans** [comment]
OK, build is fixed now that https://github.com/NixOS/nixpkgs/commit/d942d4473d1165cf9d8a1171bbad154f77239ba0 is in `nixos-unstable`, and comments have been addressed.

**Comment by berbiche** [comment]
@sumnerevans feel free to merge :+1:

**Pull request merged**
